### PR TITLE
fix: DHIS2-8910 filter out Data Element which are not present in Program Stage

### PR DESCRIPTION
**Pull request by gnespolino**

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/shared/preprocess/FilteringOutUndeclaredDataElementsProcessor.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+import org.hisp.dhis.common.IdScheme;
+import org.hisp.dhis.dxf2.events.event.Event;
+import org.hisp.dhis.dxf2.events.importer.Processor;
+import org.hisp.dhis.dxf2.events.importer.context.WorkContext;
+import org.hisp.dhis.program.ProgramStage;
+
+/**
+ * @author Giuseppe Nespolino <g.nespolino@gmail.com>
+ */
+public class FilteringOutUndeclaredDataElementsProcessor implements Processor
```
Comment (luciano-fiandesio) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/shared/preprocess/FilteringOutUndeclaredDataElementsProcessor.java:45]: Can you add a short Javadoc that explain what this Pre-proc is doing?

**Review by luciano-fiandesio — CHANGES_REQUESTED** [review]
Please, add a unit test for the new Processor. You can check  one of the other preprocessor tests for guidance (or just let me know)

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/shared/preprocess/FilteringOutUndeclaredDataElementsProcessor.java (hunk 1 above)
Comment (gnespolino) [reply dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/shared/preprocess/FilteringOutUndeclaredDataElementsProcessor.java:45]: Yes, sure

**Comment by gnespolino** [comment]
> Please, add a unit test for the new Processor. You can check one of the other preprocessor tests for guidance (or just let me know)

Done!

**Review by luciano-fiandesio — APPROVED** [review]

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/shared/preprocess/FilteringOutUndeclaredDataElementsProcessor.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+                    .filter( dataValue -> programStageDataElementUids.contains( dataValue.getDataElement() ) )
+                    .collect( Collectors.toSet() ) );
+        }
+    }
+
+    private Set<String> getDataElementUidsFromProgramStage( String programStageUid, WorkContext ctx )
+    {
+        IdScheme scheme = ctx.getImportOptions().getIdSchemes().getProgramStageIdScheme();
+        ProgramStage programStage = ctx.getProgramStage( scheme, programStageUid );
+        return Optional.ofNullable(programStage)
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/shared/preprocess/FilteringOutUndeclaredDataElementsProcessor.java:70]:
```suggestion
        return Optional.ofNullable( programStage )
```

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/shared/preprocess/FilteringOutUndeclaredDataElementsProcessor.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+                .map(ProgramStage::getDataElements)
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/shared/preprocess/FilteringOutUndeclaredDataElementsProcessor.java:71]:
```suggestion
                .map( ProgramStage::getDataElements )
```

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/shared/preprocess/FilteringOutUndeclaredDataElementsProcessor.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+                .orElse(Collections.emptySet())
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/shared/preprocess/FilteringOutUndeclaredDataElementsProcessor.java:72]:
```suggestion
                .orElse( Collections.emptySet() )
```

**Review by stian-sandvold — APPROVED** [review]

**Pull request merged**